### PR TITLE
fix chapter 4 isPrime exercise

### PR DESCRIPTION
**Pull request by bittersweetshimmer**
Prime numbers by definition are natural numbers greater than 1.

### CODE REVIEW

**Comment by hdgarrood** [comment]
I agree with fixing the tests so that they don’t consider 1 to be prime but I feel like having `factors 1` return `[[1,1]]` is a bit suspect. Perhaps `factors 1` should return an empty list?

**Comment by milesfrain** [comment]
Nice catch. Missed this one during code review.

File: exercises/chapter4/test/no-peeking/Solutions.purs [hunk 1]
```diff
@@ -41,7 +41,7 @@ keepNonNegativeRewrite :: Array Number -> Array Number
 keepNonNegativeRewrite arr = (\n -> n >= 0.0) <$?> arr
 
 isPrime :: Int -> Boolean
-isPrime n = eq 1 $ length $ factors n
+isPrime n = (n > 1) && (eq 1 $ length $ factors n)
```
Comment (milesfrain) [inline exercises/chapter4/test/no-peeking/Solutions.purs:44]:
```suggestion
isPrime n = n > 1 && length (factors n) == 1
```
If you want to save some parens. Not sure if this makes things less clear.

File: exercises/chapter4/test/Main.purs [hunk 2]
```diff
@@ -70,8 +70,10 @@ Note to reader: Delete this line to expand comment block -}
       test "Exercise - isPrime" do
```
Comment (milesfrain) [inline exercises/chapter4/test/Main.purs:70]: Should we test negatives too?

File: exercises/chapter4/test/no-peeking/Solutions.purs (hunk 1 above)
Comment (hdgarrood) [reply exercises/chapter4/test/no-peeking/Solutions.purs:44]: Yeah, this is preferable in my view. A linter (if one existed) would probably complain about what you have here.

Comment (hdgarrood) [reply exercises/chapter4/test/no-peeking/Solutions.purs:44]: Because of the redundant parens, that is

**Comment by hdgarrood** [comment]
I can’t think of a better approach than adding an `if x <= 1 then [] else ...` around the function body. I wouldn’t recommend handling negatives, as if you start handling negatives it just makes things more awkward since you have two possibilities for every factor pair. That is, should `factors (-4)` return [2, -2] or [-2, 2] (or both)? Plus, if the factors function is allowed to return negatives, should [-2, -2] count as a factor pair for 4? Accepting negatives doesn’t really allow the function to provide any new or useful information, since any factorizarion of -n is just a factorization of n with an odd number of minus signs thrown in, so I think we should ignore negatives here.

**Comment by bittersweetshimmer** [comment]
1 is a factor of 1. The root of the problem is the fact that `factors` returns pairs of factors instead of just an array. Imagine something like:

```purescript
factors :: Int -> Array Int
factors 0 = []
factors n = do 
  i <- 1 .. abs n
  guard $ n `mod` i == 0
  [i, -i]

isPrime :: Int -> Boolean
isPrime n = n > 1 && length (filter (greaterThan 0) (factors n)) == 2
```

Now `factors` works as expected, just needs 0 as a special case since every number other than 0 divides 0.

File: exercises/chapter4/test/Main.purs (hunk 2 above)
Comment (bittersweetshimmer) [reply exercises/chapter4/test/Main.purs:70]: They are in the function domain so I think it's reasonable to test them.

**Comment by hdgarrood** [comment]
I don’t agree that the fact that factors returns an array is a problem; it’s just a slightly different function from what you’re thinking of. The factors function returns the list of pairs satisfying the property that the product of each pair is the number you started with. If you just return an array then you’ve lost the information of which of those factors can multiply together to reconstruct the original number. More importantly, it’s more appropriate for demonstrating how do notation works with arrays, as you have two occurrences of <-.

Also, since we are going to go on to talk about primes, we are definitely in the realm of natural numbers. If there were a more ergonomic natural number type for PureScript, the chapter would probably be using it rather than Int here, but that doesn’t mean we can’t still approach the exercises with the perspective of natural numbers. If we decided that we wanted our factors function to instead work in the realm of integers, then our isPrime function would probably want to consider -2 to be prime (which, technically it is if you’re working with the integers, but that perspective isn’t very useful until you’ve encountered integral domains, i.e. way beyond what we should assume here).

I agree that since 1 is a factor of 1 that’s an argument in favour of having `factors 1` return [[1,1]]. I’m just not sure about it because 1 is the identity element for multiplication, so I think it’s reasonable to say that [1,1] is a useless factor pair which should be ignored. On reflection I’m not really concerned either way, though. The part I do have a stronger opinion on is that I think factors should return an empty array for negative inputs.

**Comment by bittersweetshimmer** [comment]
> The factors function returns the list of pairs satisfying the property that the product of each pair is the number you started with.

Then the name `factors` is misleading.

>  you’ve lost the information of which of those factors can multiply together to reconstruct the original number

If _a_ is a factor of _b_, then _b / a_ is also a factor of _b_, you don't lose anything, returning pairs just makes it unnecessarily convoluted.

> then our isPrime function would probably want to consider -2 to be prime (which, technically it is if you’re working with the integers

Negative numbers can't be prime, since the definition of a prime number states that they are natural numbers greater than 1, and that still applies in the superset of integers.

> because 1 is the identity element for multiplication, so I think it’s reasonable to say that [1,1] is a useless factor pair which should be ignored

For every prime `p`, `factors p` should return `[[p, 1]]` (or `[[1, p]]`), so that's still a special case for only `factors 1` (which would be false).

I'm not saying to change the whole `factors` signature, just that it's the thing that caused this problem.

I committed changes only to `isPrime` to check whether argument is greater than 1, so `factors 0`/`factors 1` never gets called anyway.

**Comment by hdgarrood** [comment]
This isn't really going anywhere, let's just merge this.

**Pull request merged**